qiskit_optimization/converters/indicator_to_inequality.py

```diff
@@ -0,0 +1,175 @@
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""The inequality to equality converter."""
+
+from typing import List, Optional, Union
+
+import numpy as np
+
+from .quadratic_program_converter import QuadraticProgramConverter
+from ..exceptions import QiskitOptimizationError
+from ..problems.constraint import Constraint
+from ..problems.quadratic_objective import QuadraticObjective
+from ..problems.quadratic_program import QuadraticProgram
+from ..problems.variable import Variable
+
+
+class IndicatorToInequality(QuadraticProgramConverter):
+    """Convert inequality constraints into equality constraints by introducing slack variables.
+
+    Examples:
+        >>> from qiskit_optimization.problems import QuadraticProgram
+        >>> from qiskit_optimization.converters import IndicatorToInequality
+        >>> problem = QuadraticProgram()
+        >>> # define a problem
+        >>> conv = IndicatorToInequality()
+        >>> problem2 = conv.convert(problem)
+    """
+
+    _delimiter = "@"  # users are supposed not to use this character in variable names
+
+    def __init__(self) -> None:
+        self._src_num_var = 0
+        self._dst = None  # type: Optional[QuadraticProgram]
+
+    def convert(self, problem: QuadraticProgram) -> QuadraticProgram:
+        """Convert a problem with indicator constraints into one with only inequality constraints.
+
+        Args:
+            problem: The problem to be solved, that may contain indicator constraints.
+
+        Returns:
+            The converted problem, that contain only inequality constraints.
+
+        Raises:
+            QiskitOptimizationError: If a variable type is not supported.
+            QiskitOptimizationError: If an unsupported mode is selected.
+            QiskitOptimizationError: If an unsupported sense is specified.
+        """
+        self._src_num_var = problem.get_num_vars()
+        self._dst = QuadraticProgram(name=problem.name)
+
+        # Copy variables
+        for x in problem.variables:
+            if x.vartype == Variable.Type.BINARY:
+                self._dst.binary_var(name=x.name)
+            elif x.vartype == Variable.Type.INTEGER:
+                self._dst.integer_var(name=x.name, lowerbound=x.lowerbound, upperbound=x.upperbound)
+            elif x.vartype == Variable.Type.CONTINUOUS:
+                self._dst.continuous_var(
+                    name=x.name, lowerbound=x.lowerbound, upperbound=x.upperbound
+                )
+            else:
+                raise QiskitOptimizationError("Unsupported variable type {}".format(x.vartype))
```

Comment:
I think generally now we have gone over to using f Strings now, since we dropped 3.5 Python support, rather than using this explicit .format() syntax. Not that its incorrect of course to do this, just that going forwards I thin it might be preferable to stick to f Strings so its all done one way.